// next-sanity/dist/studio/head/NextStudioHead.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var jsxRuntime = require('react/jsx-runtime');
var appleTouchIcon = require('./apple-touch-icon.png.cjs');
var favicon = require('./favicon.ico.cjs');
var favicon$1 = require('./favicon.svg.cjs');
const faviconPng = appleTouchIcon;
const faviconIco = favicon;
const faviconSvg = favicon$1;
function NextStudioHead(props) {
  const {
    charSet = "utf-8",
    viewport = true,
    robots = "noindex",
    referrer = "same-origin",
    favicons = true,
    title = "Sanity"
  } = props;
  return /* @__PURE__ */jsxRuntime.jsxs(jsxRuntime.Fragment, {
    children: [charSet && /* @__PURE__ */jsxRuntime.jsx("meta", {
      charSet
    }, "charset"), viewport && /* @__PURE__ */jsxRuntime.jsx("meta", {
      name: "viewport",
      content: "width=device-width,initial-scale=1,viewport-fit=cover"
    }, "viewport"), robots && /* @__PURE__ */jsxRuntime.jsx("meta", {
      name: "robots",
      content: robots
    }, "robots"), referrer && /* @__PURE__ */jsxRuntime.jsx("meta", {
      name: "referrer",
      content: referrer
    }, "referrer"), title && /* @__PURE__ */jsxRuntime.jsx("title", {
      children: title
    }), favicons && /* @__PURE__ */jsxRuntime.jsx("link", {
      rel: "icon",
      href: faviconIco,
      sizes: "any"
    }, "favicon.ico"), favicons && /* @__PURE__ */jsxRuntime.jsx("link", {
      rel: "apple-touch-icon",
      href: faviconPng
    }, "apple-touch-icon.png"), favicons && /* @__PURE__ */jsxRuntime.jsx("link", {
      rel: "icon",
      href: faviconSvg,
      type: "image/svg+xml"
    }, "favicon.svg")]
  });
}
exports.NextStudioHead = NextStudioHead;
