// next-sanity/dist/studio/head/NextStudioHead.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import _faviconPng from './apple-touch-icon.png.js';
import _faviconIco from './favicon.ico.js';
import _faviconSvg from './favicon.svg.js';
const faviconPng = _faviconPng;
const faviconIco = _faviconIco;
const faviconSvg = _faviconSvg;
function NextStudioHead(props) {
  const {
    charSet = "utf-8",
    viewport = true,
    robots = "noindex",
    referrer = "same-origin",
    favicons = true,
    title = "Sanity"
  } = props;
  return /* @__PURE__ */jsxs(Fragment, {
    children: [charSet && /* @__PURE__ */jsx("meta", {
      charSet
    }, "charset"), viewport && /* @__PURE__ */jsx("meta", {
      name: "viewport",
      content: "width=device-width,initial-scale=1,viewport-fit=cover"
    }, "viewport"), robots && /* @__PURE__ */jsx("meta", {
      name: "robots",
      content: robots
    }, "robots"), referrer && /* @__PURE__ */jsx("meta", {
      name: "referrer",
      content: referrer
    }, "referrer"), title && /* @__PURE__ */jsx("title", {
      children: title
    }), favicons && /* @__PURE__ */jsx("link", {
      rel: "icon",
      href: faviconIco,
      sizes: "any"
    }, "favicon.ico"), favicons && /* @__PURE__ */jsx("link", {
      rel: "apple-touch-icon",
      href: faviconPng
    }, "apple-touch-icon.png"), favicons && /* @__PURE__ */jsx("link", {
      rel: "icon",
      href: faviconSvg,
      type: "image/svg+xml"
    }, "favicon.svg")]
  });
}
export { NextStudioHead };
