// @sanity/preview-kit/dist/GroqStoreProvider/GroqStoreProvider.cjs
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var DefaultEventSource = require('@sanity/eventsource');
var groqStore = require('@sanity/groq-store');
var react = require('react');
var context = require('../context.cjs');
var utils = require('../utils.cjs');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var DefaultEventSource__default = /*#__PURE__*/_interopDefaultCompat(DefaultEventSource);
const GroqStoreProviderInternal = react.memo(function GroqStoreProvider(props) {
  const {
    children,
    logger,
    // The rest is the store config
    ...config
  } = props;
  const [ready] = react.useState(() => /* @__PURE__ */new Set());
  const [loadedListenersContext, updateLoadedListeners] = utils.useLoadingListenersContext(ready);
  const [snapshots] = react.useState(() => /* @__PURE__ */new Map());
  const [store] = react.useState(() => groqStore.groqStore({
    // Override some of the store defaults
    EventSource: props.token ? DefaultEventSource__default.default : void 0,
    listen: true,
    overlayDrafts: true,
    documentLimit: utils.DEFAULT_MAX_DOCUMENTS,
    // Spread in the rest
    ...config
  }));
  const report = react.useMemo(() => {
    if (config.listen) {
      return "Updates are applied in real-time. The cache is set to max ".concat(config.documentLimit || utils.DEFAULT_MAX_DOCUMENTS, " documents.");
    }
    return "Updates require a manual refresh. The cache is set to max ".concat(config.documentLimit || utils.DEFAULT_MAX_DOCUMENTS, " documents.");
  }, [config.documentLimit, config.listen]);
  react.useEffect(() => {
    if (logger) {
      logger.log("[@sanity/preview-kit]: With the current configuration you can expect that: ".concat(report));
    }
  }, [logger, report]);
  const [error, setError] = react.useState(null);
  if (error) throw error;
  const [listenerContext] = react.useState(() => {
    return function defineListener(initialSnapshot, query, params) {
      const key = utils.getQueryCacheKey(query, params);
      if (!snapshots.has(key)) {
        snapshots.set(key, initialSnapshot);
      }
      const subscribe = onStoreChange => {
        if (!ready.has(key)) {
          store.query(query, params).then(result => {
            if (!ready.has(key)) {
              snapshots.set(key, result);
              ready.add(key);
              updateLoadedListeners();
              onStoreChange();
            }
          }, setError);
        }
        if (!config.listen) {
          return () => {};
        }
        const subscription = store.subscribe(query, params, (err, result) => {
          if (err) {
            setError(err);
          } else if (ready.has(key)) {
            snapshots.set(key, result);
            onStoreChange();
          }
        });
        return () => subscription.unsubscribe();
      };
      const getSnapshot = () => snapshots.get(key);
      return {
        subscribe,
        getSnapshot
      };
    };
  });
  return /* @__PURE__ */jsxRuntime.jsx(context.defineListenerContext.Provider, {
    value: listenerContext,
    children: /* @__PURE__ */jsxRuntime.jsx(context.LoadedListenersContext.Provider, {
      value: loadedListenersContext,
      children: /* @__PURE__ */jsxRuntime.jsx(context.IsEnabledContext.Provider, {
        value: true,
        children
      })
    })
  });
});
const GroqStoreProvider2 = react.memo(function GroqStoreProvider3(props) {
  var _a;
  const {
    children,
    client,
    cache,
    logger
  } = props;
  const {
    projectId,
    dataset,
    token,
    // eslint-disable-next-line no-warning-comments
    // @TODO @sanity/groq-store should handle `perspective` directly
    perspective = "previewDrafts",
    requestTagPrefix
  } = react.useMemo(() => client.config(), [client]);
  return /* @__PURE__ */jsxRuntime.jsx(GroqStoreProviderInternal, {
    projectId,
    dataset,
    token,
    logger,
    listen: (_a = cache == null ? void 0 : cache.listen) != null ? _a : true,
    documentLimit: cache == null ? void 0 : cache.maxDocuments,
    overlayDrafts: perspective === "previewDrafts",
    includeTypes: cache == null ? void 0 : cache.includeTypes,
    requestTagPrefix,
    children
  });
});
GroqStoreProvider2.displayName = "GroqStoreProvider";
var GroqStoreProvider2$1 = GroqStoreProvider2;
module.exports = GroqStoreProvider2$1;
