// sanity-plugin-vercel-deploy/dist/index.esm.js
import { useClient as useClient$1, FormField, useColorScheme, definePlugin } from 'sanity';
import { route } from 'sanity/router';
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import axios from 'axios';
import { nanoid } from 'nanoid';
import { useState, useEffect } from 'react';
import { Flex, Badge, Spinner, Box, Text, Card, Label, Stack, Inline, Tooltip, Avatar, useToast, Heading, Code, Button, MenuButton, Menu, MenuItem, Dialog, Grid, TextInput, Switch, ThemeProvider, studioTheme, ToastProvider, Container } from '@sanity/ui';
import spacetime from 'spacetime';
import useSWR from 'swr';
import { TransferIcon, EllipsisVerticalIcon, ClockIcon, EditIcon, TrashIcon } from '@sanity/icons';
const deployIcon = () => {
  return /* @__PURE__ */jsxs("svg", {
    "data-sanity-icon": "true",
    viewBox: "0 0 25 25",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg",
    preserveAspectRatio: "xMidYMid",
    width: "1em",
    height: "1em",
    children: [/* @__PURE__ */jsx("path", {
      d: "M10.06,19.53H4.27L12.5,5.21l8.23,14.32H14.94",
      style: {
        stroke: "currentColor",
        strokeWidth: "1.2"
      }
    }), /* @__PURE__ */jsx("path", {
      d: "M12.5,12.58v7.51",
      style: {
        stroke: "currentColor",
        strokeWidth: "1.2"
      }
    }), /* @__PURE__ */jsx("path", {
      d: "M15.12,16.76,12.5,12.58,10,16.71",
      style: {
        stroke: "currentColor",
        strokeWidth: "1.2"
      }
    })]
  });
};
const DeployStatus = _ref => {
  let {
    status,
    justify,
    children
  } = _ref;
  const titleCase = str => {
    return str.toLowerCase().split(" ").map(word => {
      return word.charAt(0).toUpperCase() + word.slice(1);
    }).join(" ");
  };
  const badgeTone = {
    LOADING: "default",
    ERROR: "critical",
    INITIATED: "default",
    CANCELED: "default",
    READY: "positive",
    BUILDING: "caution",
    QUEUED: "default"
  }[status] || "default";
  const badgeMode = {
    LOADING: "outline",
    READY: "outline",
    CANCELED: "outline"
  }[status] || "default";
  return /* @__PURE__ */jsxs(Flex, {
    wrap: "nowrap",
    align: "center",
    justify,
    children: [/* @__PURE__ */jsx(Badge, {
      mode: badgeMode,
      tone: badgeTone,
      padding: 2,
      fontSize: 1,
      children: titleCase(status)
    }), children]
  });
};
const DeployHistory = _ref2 => {
  let {
    url,
    vercelProject,
    vercelToken,
    vercelTeam
  } = _ref2;
  var _a;
  const [deployments, setDeployments] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const deployHookId = (_a = url == null ? void 0 : url.split("/").pop()) == null ? void 0 : _a.split("?").shift();
  useEffect(() => {
    if (!vercelProject) {
      return;
    }
    setLoading(true);
    axios.get("https://api.vercel.com/v5/now/deployments?projectId=".concat(vercelProject, "&meta-deployHookId=").concat(deployHookId, "&limit=6").concat((vercelTeam == null ? void 0 : vercelTeam.id) ? "&teamId=".concat(vercelTeam == null ? void 0 : vercelTeam.id) : ""), {
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer ".concat(vercelToken)
      }
    }).then(_ref3 => {
      let {
        data
      } = _ref3;
      setDeployments(data.deployments);
      setLoading(false);
      setError(false);
    }).catch(e => {
      setLoading(false);
      setError(true);
      console.warn(e);
    });
  }, [url, vercelProject, vercelTeam == null ? void 0 : vercelTeam.id, vercelToken]);
  if (loading) {
    return /* @__PURE__ */jsxs(Flex, {
      direction: "column",
      align: "center",
      justify: "center",
      paddingTop: 3,
      children: [/* @__PURE__ */jsx(Spinner, {
        size: 4
      }), /* @__PURE__ */jsx(Box, {
        padding: 4,
        children: /* @__PURE__ */jsx(Text, {
          size: 2,
          children: "loading deployment history..."
        })
      })]
    });
  }
  if (error) {
    return /* @__PURE__ */jsx(Card, {
      padding: 4,
      radius: 2,
      shadow: 1,
      tone: "critical",
      children: /* @__PURE__ */jsxs(Text, {
        size: 2,
        align: "center",
        children: ["Could not load deployments for ", vercelProject]
      })
    });
  }
  return /* @__PURE__ */jsxs(Box, {
    as: "ul",
    padding: 0,
    children: [/* @__PURE__ */jsx(Card, {
      as: "li",
      padding: 4,
      borderBottom: true,
      children: /* @__PURE__ */jsxs(Flex, {
        children: [/* @__PURE__ */jsx(Box, {
          flex: 3,
          children: /* @__PURE__ */jsx(Label, {
            muted: true,
            children: "Preview URL"
          })
        }), /* @__PURE__ */jsx(Box, {
          flex: 1,
          marginLeft: 2,
          children: /* @__PURE__ */jsx(Label, {
            muted: true,
            children: "State"
          })
        }), /* @__PURE__ */jsx(Box, {
          flex: 3,
          marginLeft: 2,
          style: {
            maxWidth: "40%"
          },
          children: /* @__PURE__ */jsx(Label, {
            muted: true,
            children: "Commit"
          })
        }), /* @__PURE__ */jsx(Box, {
          flex: 2,
          marginLeft: 2,
          children: /* @__PURE__ */jsx(Label, {
            align: "right",
            muted: true,
            children: "Deployed At"
          })
        })]
      })
    }), deployments == null ? void 0 : deployments.map(deployment => {
      var _a2, _b, _c, _d, _e;
      return /* @__PURE__ */jsx(Card, {
        as: "li",
        padding: 4,
        borderBottom: true,
        children: /* @__PURE__ */jsxs(Flex, {
          align: "center",
          children: [/* @__PURE__ */jsx(Box, {
            flex: 3,
            children: /* @__PURE__ */jsx(Text, {
              weight: "semibold",
              children: /* @__PURE__ */jsx(Box, {
                style: {
                  whiteSpace: "nowrap",
                  overflow: "hidden",
                  textOverflow: "ellipsis"
                },
                children: /* @__PURE__ */jsx("a", {
                  href: "https://".concat(deployment.url),
                  target: "_blank",
                  rel: "noopener noreferrer",
                  style: {
                    color: "inherit"
                  },
                  children: deployment.url
                })
              })
            })
          }), /* @__PURE__ */jsx(Box, {
            flex: 1,
            marginLeft: 2,
            children: /* @__PURE__ */jsx(Text, {
              children: /* @__PURE__ */jsx(DeployStatus, {
                status: deployment.state
              })
            })
          }), /* @__PURE__ */jsx(Box, {
            flex: 3,
            marginLeft: 2,
            style: {
              maxWidth: "40%"
            },
            children: /* @__PURE__ */jsxs(Stack, {
              space: 2,
              children: [/* @__PURE__ */jsx(Text, {
                children: /* @__PURE__ */jsx(Box, {
                  style: {
                    whiteSpace: "nowrap",
                    overflow: "hidden",
                    textOverflow: "ellipsis"
                  },
                  children: (_a2 = deployment.meta) == null ? void 0 : _a2.githubCommitMessage
                })
              }), /* @__PURE__ */jsx(Text, {
                size: 2,
                muted: true,
                children: /* @__PURE__ */jsxs(Inline, {
                  space: 3,
                  children: [/* @__PURE__ */jsx(TransferIcon, {}), (_b = deployment.meta) == null ? void 0 : _b.githubCommitRef]
                })
              })]
            })
          }), /* @__PURE__ */jsx(Flex, {
            flex: 2,
            justify: "flex-end",
            marginLeft: 2,
            children: /* @__PURE__ */jsxs(Inline, {
              space: 2,
              children: [/* @__PURE__ */jsx(Text, {
                style: {
                  whiteSpace: "nowrap"
                },
                muted: true,
                children: spacetime.now().since(spacetime(deployment.created)).rounded
              }), /* @__PURE__ */jsx(Tooltip, {
                content: /* @__PURE__ */jsx(Box, {
                  padding: 2,
                  children: /* @__PURE__ */jsx(Text, {
                    muted: true,
                    size: 1,
                    children: (_c = deployment.creator) == null ? void 0 : _c.username
                  })
                }),
                fallbackPlacements: ["right", "left"],
                placement: "top",
                children: /* @__PURE__ */jsx(Avatar, {
                  alt: (_d = deployment.creator) == null ? void 0 : _d.username,
                  src: "https://vercel.com/api/www/avatar/".concat((_e = deployment.creator) == null ? void 0 : _e.uid, "?&s=48"),
                  size: 1
                })
              })]
            })
          })]
        })
      }, deployment.uid);
    })]
  });
};
const useClient = () => {
  return useClient$1({
    apiVersion: "2022-09-14"
  });
};
const fetcher = (url, token) => axios.get(url, {
  headers: {
    "content-type": "application/json",
    Authorization: "Bearer ".concat(token)
  }
}).then(res => res.data);
const initialDeploy$1 = {
  title: "",
  project: "",
  team: "",
  url: "",
  token: "",
  disableDeleteAction: false
};
const DeployItem = _ref4 => {
  let {
    name,
    url,
    _id,
    vercelProject,
    vercelToken,
    vercelTeam,
    disableDeleteAction
  } = _ref4;
  var _a;
  const client = useClient();
  const [isLoading, setIsLoading] = useState(true);
  const [isDeploying, setDeploying] = useState(false);
  const [isHistoryOpen, setIsHistoryOpen] = useState(false);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [isFormOpen, setIsFormOpen] = useState(false);
  const [pendingDeploy, setpendingDeploy] = useState(initialDeploy$1);
  const [errorMessage, setErrorMessage] = useState(null);
  const [status, setStatus] = useState("LOADING");
  const [timestamp, setTimestamp] = useState(null);
  const [buildTime, setBuildTime] = useState(null);
  const toast = useToast();
  const deployHookId = (_a = url == null ? void 0 : url.split("/").pop()) == null ? void 0 : _a.split("?").shift();
  const {
    data: projectData
  } = useSWR(["https://api.vercel.com/v8/projects/".concat(vercelProject).concat((vercelTeam == null ? void 0 : vercelTeam.id) ? "?teamId=".concat(vercelTeam == null ? void 0 : vercelTeam.id) : ""), vercelToken], (path, token) => fetcher(path, token), {
    errorRetryCount: 3,
    onError: err => {
      var _a2, _b, _c;
      setStatus("ERROR");
      setErrorMessage((_c = (_b = (_a2 = err.response) == null ? void 0 : _a2.data) == null ? void 0 : _b.error) == null ? void 0 : _c.message);
      setIsLoading(false);
    }
  });
  const {
    data: deploymentData
  } = useSWR(() => ["https://api.vercel.com/v5/now/deployments?projectId=".concat(projectData.id, "&meta-deployHookId=").concat(deployHookId, "&limit=1").concat((vercelTeam == null ? void 0 : vercelTeam.id) ? "&teamId=".concat(vercelTeam == null ? void 0 : vercelTeam.id) : ""), vercelToken], (path, token) => fetcher(path, token), {
    errorRetryCount: 3,
    refreshInterval: isDeploying ? 5e3 : 0,
    onError: err => {
      var _a2, _b, _c;
      setStatus("ERROR");
      setErrorMessage((_c = (_b = (_a2 = err.response) == null ? void 0 : _a2.data) == null ? void 0 : _b.error) == null ? void 0 : _c.message);
      setIsLoading(false);
    }
  });
  const onDeploy = (_name, _url) => {
    setStatus("INITIATED");
    setDeploying(true);
    setTimestamp(null);
    setBuildTime(null);
    axios.post(url).then(() => {
      toast.push({
        status: "success",
        title: "Success!",
        description: "Triggered Deployment: ".concat(_name)
      });
    }).catch(err => {
      setDeploying(false);
      toast.push({
        status: "error",
        title: "Deploy Failed.",
        description: "".concat(err)
      });
    });
  };
  const onCancel = (id, token) => {
    setIsLoading(true);
    axios.patch("https://api.vercel.com/v12/deployments/".concat(id, "/cancel"), null, {
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer ".concat(token)
      },
      params: {
        ...(vercelTeam ? {
          teamId: vercelTeam == null ? void 0 : vercelTeam.id
        } : {})
      }
    }).then(res => res.data).then(res => {
      setStatus("CANCELED");
      setDeploying(false);
      setIsLoading(false);
      setBuildTime(null);
      setTimestamp(res.canceledAt);
    });
  };
  const onRemove = (_name, id) => {
    setIsLoading(true);
    client.delete(id).then(() => {
      toast.push({
        status: "success",
        title: "Successfully deleted deployment: ".concat(_name)
      });
    });
  };
  const onEdit = () => {
    setpendingDeploy({
      title: name,
      project: vercelProject,
      team: vercelTeam == null ? void 0 : vercelTeam.slug,
      url,
      token: vercelToken,
      disableDeleteAction
    });
    setIsFormOpen(true);
  };
  const onSubmitEdit = async () => {
    var _a2;
    let vercelTeamID;
    let vercelTeamName;
    setIsSubmitting(true);
    if (pendingDeploy.team) {
      try {
        const fetchTeam = await axios.get("https://api.vercel.com/v2/teams?slug=".concat(pendingDeploy.team), {
          headers: {
            Authorization: "Bearer ".concat(pendingDeploy.token)
          }
        });
        if (!((_a2 = fetchTeam == null ? void 0 : fetchTeam.data) == null ? void 0 : _a2.id)) {
          throw new Error("No team id found");
        }
        vercelTeamID = fetchTeam.data.id;
        vercelTeamName = fetchTeam.data.name;
      } catch (error) {
        console.error(error);
        setIsSubmitting(false);
        toast.push({
          status: "error",
          title: "No Team found!",
          closable: true,
          description: "Make sure the token you provided is valid and that the team\u2019s slug correspond to the one you see in Vercel"
        });
        return;
      }
    }
    client.patch(_id).set({
      name: pendingDeploy.title,
      url: pendingDeploy.url,
      vercelProject: pendingDeploy.project,
      vercelTeam: {
        slug: pendingDeploy.team || void 0,
        name: vercelTeamName || void 0,
        id: vercelTeamID || void 0
      },
      vercelToken: pendingDeploy.token,
      disableDeleteAction: pendingDeploy.disableDeleteAction
    }).commit().then(() => {
      toast.push({
        status: "success",
        title: "Success!",
        description: "Updated Deployment: ".concat(pendingDeploy.title)
      });
      setIsFormOpen(false);
      setIsSubmitting(false);
    });
  };
  useEffect(() => {
    let isSubscribed = true;
    if ((deploymentData == null ? void 0 : deploymentData.deployments) && isSubscribed) {
      const latestDeployment = deploymentData.deployments[0];
      setStatus((latestDeployment == null ? void 0 : latestDeployment.state) || "READY");
      if (latestDeployment == null ? void 0 : latestDeployment.created) {
        setTimestamp(latestDeployment == null ? void 0 : latestDeployment.created);
      }
      setIsLoading(false);
    }
    return () => {
      isSubscribed = false;
    };
  }, [deploymentData]);
  useEffect(() => {
    let isSubscribed = true;
    if (isSubscribed) {
      if (status === "READY" || status === "ERROR" || status === "CANCELED") {
        setDeploying(false);
      } else if (status === "BUILDING" || status === "INITIATED") {
        setDeploying(true);
      }
    }
    return () => {
      isSubscribed = false;
    };
  }, [status]);
  const tick = _timestamp => {
    if (_timestamp) {
      setBuildTime(spacetime.now().since(spacetime(_timestamp)).rounded);
    }
  };
  useEffect(() => {
    let isTicking = true;
    const timer = setInterval(() => {
      if (isTicking && isDeploying) {
        tick(timestamp);
      }
    }, 1e3);
    if (!isDeploying) {
      clearInterval(timer);
    }
    return () => {
      isTicking = false;
      clearInterval(timer);
    };
  }, [timestamp, isDeploying]);
  return /* @__PURE__ */jsxs(Fragment, {
    children: [/* @__PURE__ */jsxs(Flex, {
      wrap: "wrap",
      direction: ["column", "column", "row"],
      align: ["flex-end", "flex-end", "center"],
      children: [/* @__PURE__ */jsx(Box, {
        flex: [4, 1],
        paddingBottom: [4, 4, 1],
        children: /* @__PURE__ */jsxs(Stack, {
          space: 3,
          children: [/* @__PURE__ */jsxs(Inline, {
            space: 2,
            children: [/* @__PURE__ */jsx(Heading, {
              as: "h2",
              size: 1,
              children: /* @__PURE__ */jsx(Text, {
                weight: "semibold",
                children: name
              })
            }), /* @__PURE__ */jsx(Badge, {
              tone: "primary",
              paddingX: 3,
              paddingY: 2,
              radius: 6,
              fontSize: 0,
              children: vercelProject
            }), (vercelTeam == null ? void 0 : vercelTeam.id) && /* @__PURE__ */jsx(Badge, {
              mode: "outline",
              paddingX: 3,
              paddingY: 2,
              radius: 6,
              fontSize: 0,
              children: vercelTeam == null ? void 0 : vercelTeam.name
            })]
          }), /* @__PURE__ */jsx(Code, {
            size: 1,
            children: /* @__PURE__ */jsx(Box, {
              style: {
                whiteSpace: "nowrap",
                overflow: "hidden",
                textOverflow: "ellipsis"
              },
              children: url
            })
          })]
        })
      }), /* @__PURE__ */jsx(Flex, {
        wrap: "nowrap",
        align: "center",
        marginLeft: [0, 0, 4],
        flex: [1, "none"],
        children: /* @__PURE__ */jsxs(Inline, {
          space: 2,
          children: [vercelToken && vercelProject && /* @__PURE__ */jsx(Box, {
            marginRight: 2,
            children: /* @__PURE__ */jsxs(Stack, {
              space: 2,
              children: [/* @__PURE__ */jsx(DeployStatus, {
                status,
                justify: "flex-end",
                children: errorMessage && /* @__PURE__ */jsx(Box, {
                  marginLeft: 2,
                  children: /* @__PURE__ */jsx(Tooltip, {
                    content: /* @__PURE__ */jsx(Box, {
                      padding: 2,
                      children: /* @__PURE__ */jsx(Text, {
                        muted: true,
                        size: 1,
                        children: errorMessage
                      })
                    }),
                    placement: "top",
                    children: /* @__PURE__ */jsx(Badge, {
                      mode: "outline",
                      tone: "critical",
                      children: "?"
                    })
                  })
                })
              }), /* @__PURE__ */jsx(Text, {
                align: "right",
                size: 1,
                muted: true,
                children: isDeploying ? buildTime || "--" : timestamp ? spacetime.now().since(spacetime(timestamp)).rounded : "--"
              })]
            })
          }), /* @__PURE__ */jsx(Button, {
            type: "button",
            tone: "positive",
            disabled: isDeploying || isLoading,
            loading: isDeploying || isLoading,
            onClick: () => onDeploy(name),
            paddingX: [5],
            paddingY: [4],
            radius: 3,
            text: "Deploy"
          }), isDeploying && (status === "BUILDING" || status === "QUEUED") && /* @__PURE__ */jsx(Button, {
            type: "button",
            tone: "critical",
            onClick: () => {
              onCancel(deploymentData.deployments[0].uid, vercelToken);
            },
            radius: 3,
            text: "Cancel"
          }), /* @__PURE__ */jsx(MenuButton, {
            id: _id,
            button: /* @__PURE__ */jsx(Button, {
              mode: "bleed",
              icon: EllipsisVerticalIcon,
              disabled: isDeploying || isLoading
            }),
            popover: {
              portal: true,
              placement: "bottom-end"
            },
            menu: /* @__PURE__ */jsxs(Menu, {
              children: [/* @__PURE__ */jsx(MenuItem, {
                text: "History",
                icon: ClockIcon,
                onClick: () => setIsHistoryOpen(true),
                disabled: !(deploymentData == null ? void 0 : deploymentData.deployments.length)
              }), /* @__PURE__ */jsx(MenuItem, {
                text: "Edit",
                icon: EditIcon,
                tone: "primary",
                onClick: () => onEdit()
              }), !disableDeleteAction && /* @__PURE__ */jsx(MenuItem, {
                text: "Delete",
                icon: TrashIcon,
                tone: "critical",
                onClick: () => onRemove(name, _id)
              })]
            })
          })]
        })
      })]
    }), isFormOpen && /* @__PURE__ */jsx(Dialog, {
      header: "Edit Project Deployment",
      id: "update-webhook",
      width: 1,
      onClickOutside: () => setIsFormOpen(false),
      onClose: () => setIsFormOpen(false),
      footer: /* @__PURE__ */jsx(Box, {
        padding: 3,
        children: /* @__PURE__ */jsxs(Grid, {
          columns: 2,
          gap: 3,
          children: [/* @__PURE__ */jsx(Button, {
            padding: 4,
            mode: "ghost",
            text: "Cancel",
            onClick: () => setIsFormOpen(false)
          }), /* @__PURE__ */jsx(Button, {
            padding: 4,
            text: "Update",
            tone: "primary",
            loading: isSubmitting,
            onClick: () => onSubmitEdit(),
            disabled: isSubmitting || !pendingDeploy.project || !pendingDeploy.url || !pendingDeploy.token
          })]
        })
      }),
      children: /* @__PURE__ */jsx(Box, {
        padding: 4,
        children: /* @__PURE__ */jsxs(Stack, {
          space: 4,
          children: [/* @__PURE__ */jsx(FormField, {
            title: "Display Title (internal use only)",
            description: /* @__PURE__ */jsxs(Fragment, {
              children: ["This should be the environment you are deploying to, like", " ", /* @__PURE__ */jsx("em", {
                children: "Production"
              }), " or ", /* @__PURE__ */jsx("em", {
                children: "Staging"
              })]
            }),
            children: /* @__PURE__ */jsx(TextInput, {
              type: "text",
              value: pendingDeploy.title,
              onChange: e => {
                e.persist();
                const title = e.target.value;
                setpendingDeploy(prevState => ({
                  ...prevState,
                  ...{
                    title
                  }
                }));
              }
            })
          }), /* @__PURE__ */jsx(FormField, {
            title: "Vercel Project Name",
            description: "Vercel Project: Settings \u2192 General \u2192 \"Project Name\"",
            children: /* @__PURE__ */jsx(TextInput, {
              type: "text",
              value: pendingDeploy.project,
              onChange: e => {
                e.persist();
                const project = e.target.value;
                setpendingDeploy(prevState => ({
                  ...prevState,
                  ...{
                    project
                  }
                }));
              }
            })
          }), /* @__PURE__ */jsx(FormField, {
            title: "Vercel Team Name",
            description: "Required for projects under a Vercel Team: Settings \u2192 General \u2192 \"Team Name\"",
            children: /* @__PURE__ */jsx(TextInput, {
              type: "text",
              value: pendingDeploy.team,
              onChange: e => {
                e.persist();
                const team = e.target.value;
                setpendingDeploy(prevState => ({
                  ...prevState,
                  ...{
                    team
                  }
                }));
              }
            })
          }), /* @__PURE__ */jsx(FormField, {
            title: "Deploy Hook URL",
            description: "Vercel Project: Settings \u2192 Git \u2192 \"Deploy Hooks\"",
            children: /* @__PURE__ */jsx(TextInput, {
              type: "text",
              inputMode: "url",
              value: pendingDeploy.url,
              onChange: e => {
                e.persist();
                const pendingDeployUrl = e.target.value;
                setpendingDeploy(prevState => ({
                  ...prevState,
                  ...{
                    url: pendingDeployUrl
                  }
                }));
              }
            })
          }), /* @__PURE__ */jsx(FormField, {
            children: /* @__PURE__ */jsx(Card, {
              paddingY: 3,
              children: /* @__PURE__ */jsxs(Flex, {
                align: "center",
                children: [/* @__PURE__ */jsx(Switch, {
                  id: "disableDeleteAction",
                  style: {
                    display: "block"
                  },
                  onChange: e => {
                    e.persist();
                    const isChecked = e.target.checked;
                    setpendingDeploy(prevState => ({
                      ...prevState,
                      ...{
                        disableDeleteAction: isChecked
                      }
                    }));
                  },
                  checked: pendingDeploy.disableDeleteAction
                }), /* @__PURE__ */jsx(Box, {
                  flex: 1,
                  paddingLeft: 3,
                  children: /* @__PURE__ */jsx(Text, {
                    children: /* @__PURE__ */jsx("label", {
                      htmlFor: "disableDeleteAction",
                      children: 'Disable the "Delete" action for this item in production?'
                    })
                  })
                })]
              })
            })
          })]
        })
      })
    }), isHistoryOpen && /* @__PURE__ */jsx(Dialog, {
      id: "deploy-history",
      header: "Deployment History: ".concat(name),
      onClickOutside: () => setIsHistoryOpen(false),
      onClose: () => setIsHistoryOpen(false),
      width: 2,
      children: /* @__PURE__ */jsx(DeployHistory, {
        url,
        vercelProject: projectData.id,
        vercelToken,
        vercelTeam
      })
    })]
  });
};
const initialDeploy = {
  title: "",
  project: "",
  team: "",
  url: "",
  token: "",
  disableDeleteAction: false
};
const VercelDeploy = () => {
  const WEBHOOK_TYPE = "webhook_deploy";
  const WEBHOOK_QUERY = "*[_type == \"".concat(WEBHOOK_TYPE, "\"] | order(_createdAt)");
  const client = useClient();
  const {
    scheme
  } = useColorScheme();
  const [isLoading, setIsLoading] = useState(true);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [isFormOpen, setIsFormOpen] = useState(false);
  const [deploys, setDeploys] = useState([]);
  const [pendingDeploy, setpendingDeploy] = useState(initialDeploy);
  const toast = useToast();
  const onSubmit = async () => {
    var _a;
    let vercelTeamID;
    let vercelTeamName;
    setIsSubmitting(true);
    if (pendingDeploy.team) {
      try {
        const fetchTeam = await axios.get("https://api.vercel.com/v2/teams?slug=".concat(pendingDeploy.team), {
          headers: {
            Authorization: "Bearer ".concat(pendingDeploy.token)
          }
        });
        if (!((_a = fetchTeam == null ? void 0 : fetchTeam.data) == null ? void 0 : _a.id)) {
          throw new Error("No team id found");
        }
        vercelTeamID = fetchTeam.data.id;
        vercelTeamName = fetchTeam.data.name;
      } catch (error) {
        console.error(error);
        setIsSubmitting(false);
        toast.push({
          status: "error",
          title: "No Team found!",
          closable: true,
          description: "Make sure the token you provided is valid and that the team\u2019s slug correspond to the one you see in Vercel"
        });
        return;
      }
    }
    client.create({
      // Explicitly define an _id inside the vercel-deploy path to make sure it's not publicly accessible
      // This will protect users' tokens & project info. Read more: https://www.sanity.io/docs/ids
      _id: "vercel-deploy.".concat(nanoid()),
      _type: WEBHOOK_TYPE,
      name: pendingDeploy.title,
      url: pendingDeploy.url,
      vercelProject: pendingDeploy.project,
      vercelTeam: {
        slug: pendingDeploy.team || void 0,
        name: vercelTeamName || void 0,
        id: vercelTeamID || void 0
      },
      vercelToken: pendingDeploy.token,
      disableDeleteAction: pendingDeploy.disableDeleteAction
    }).then(() => {
      toast.push({
        status: "success",
        title: "Success!",
        description: "Created Deployment: ".concat(pendingDeploy.title)
      });
      setIsFormOpen(false);
      setIsSubmitting(false);
      setpendingDeploy(initialDeploy);
    });
  };
  useEffect(() => {
    let webhookSubscription;
    client.fetch(WEBHOOK_QUERY).then(w => {
      setDeploys(w);
      setIsLoading(false);
      webhookSubscription = client.listen(WEBHOOK_QUERY, {}, {
        includeResult: true
      }).subscribe({
        next: res => {
          if (res.type === "mutation") {
            const wasCreated = res.mutations.some(item => Object.prototype.hasOwnProperty.call(item, "create"));
            const wasPatched = res.mutations.some(item => Object.prototype.hasOwnProperty.call(item, "patch"));
            const wasDeleted = res.mutations.some(item => Object.prototype.hasOwnProperty.call(item, "delete"));
            const filterDeploy = deploy => deploy._id !== res.documentId;
            const updateDeploy = deploy => deploy._id === res.documentId ? res.result : deploy;
            if (wasCreated) {
              setDeploys(prevState => {
                if (res.result) {
                  return [...prevState, res.result];
                }
                return prevState;
              });
            }
            if (wasPatched) {
              setDeploys(prevState => {
                const updatedDeploys = prevState.map(updateDeploy);
                return updatedDeploys;
              });
            }
            if (wasDeleted) {
              setDeploys(prevState => prevState.filter(filterDeploy));
            }
          }
        }
      });
    });
    return () => {
      if (webhookSubscription) {
        webhookSubscription.unsubscribe();
      }
    };
  }, [WEBHOOK_QUERY, client]);
  return /* @__PURE__ */jsx(ThemeProvider, {
    theme: studioTheme,
    children: /* @__PURE__ */jsxs(ToastProvider, {
      children: [/* @__PURE__ */jsx(Container, {
        display: "grid",
        width: 6,
        style: {
          minHeight: "100%"
        },
        children: /* @__PURE__ */jsxs(Flex, {
          direction: "column",
          children: [/* @__PURE__ */jsx(Card, {
            padding: 4,
            borderBottom: true,
            children: /* @__PURE__ */jsxs(Flex, {
              align: "center",
              children: [/* @__PURE__ */jsxs(Flex, {
                flex: 1,
                align: "center",
                children: [/* @__PURE__ */jsx(Card, {
                  children: /* @__PURE__ */jsx("svg", {
                    fill: "currentColor",
                    viewBox: "0 0 512 512",
                    height: "2rem",
                    width: "2rem",
                    xmlns: "http://www.w3.org/2000/svg",
                    children: /* @__PURE__ */jsx("path", {
                      d: "M256 48l240 416H16z"
                    })
                  })
                }), /* @__PURE__ */jsx(Card, {
                  marginX: 1,
                  style: {
                    opacity: 0.15
                  },
                  children: /* @__PURE__ */jsx("svg", {
                    viewBox: "0 0 24 24",
                    width: "32",
                    height: "32",
                    stroke: "currentColor",
                    strokeWidth: "1",
                    strokeLinecap: "round",
                    strokeLinejoin: "round",
                    fill: "none",
                    shapeRendering: "geometricPrecision",
                    children: /* @__PURE__ */jsx("path", {
                      d: "M16.88 3.549L7.12 20.451"
                    })
                  })
                }), /* @__PURE__ */jsx(Card, {
                  children: /* @__PURE__ */jsx(Text, {
                    as: "h1",
                    size: 2,
                    weight: "semibold",
                    children: "Vercel Deployments"
                  })
                })]
              }), /* @__PURE__ */jsx(Box, {
                children: /* @__PURE__ */jsx(Button, {
                  type: "button",
                  fontSize: 2,
                  tone: "primary",
                  padding: 3,
                  radius: 3,
                  text: "Add Project",
                  onClick: () => setIsFormOpen(true)
                })
              })]
            })
          }), /* @__PURE__ */jsx(Card, {
            flex: 1,
            children: /* @__PURE__ */jsx(Stack, {
              as: "ul",
              children: isLoading ? /* @__PURE__ */jsx(Card, {
                as: "li",
                padding: 4,
                children: /* @__PURE__ */jsxs(Flex, {
                  direction: "column",
                  align: "center",
                  justify: "center",
                  paddingTop: 3,
                  children: [/* @__PURE__ */jsx(Spinner, {
                    size: 4
                  }), /* @__PURE__ */jsx(Box, {
                    padding: 4,
                    children: /* @__PURE__ */jsx(Text, {
                      size: 2,
                      children: "loading your deployments..."
                    })
                  })]
                })
              }) : deploys.length ? deploys.map(deploy => /* @__PURE__ */jsx(Card, {
                as: "li",
                padding: 4,
                borderBottom: true,
                children: /* @__PURE__ */jsx(DeployItem, {
                  name: deploy.name,
                  url: deploy.url,
                  _id: deploy._id,
                  vercelProject: deploy.vercelProject,
                  vercelTeam: deploy.vercelTeam,
                  vercelToken: deploy.vercelToken,
                  disableDeleteAction: deploy.disableDeleteAction
                }, deploy._id)
              }, deploy._id)) : /* @__PURE__ */jsx(Card, {
                as: "li",
                padding: 5,
                paddingTop: 6,
                children: /* @__PURE__ */jsxs(Flex, {
                  direction: "column",
                  align: "center",
                  justify: "center",
                  children: [/* @__PURE__ */jsxs("svg", {
                    xmlns: "http://www.w3.org/2000/svg",
                    fill: "none",
                    width: "150",
                    viewBox: "0 0 260 235",
                    children: [/* @__PURE__ */jsx("path", {
                      fill: scheme === "dark" ? "transparent" : "white",
                      fillRule: "evenodd",
                      stroke: scheme === "dark" ? "white" : "black",
                      strokeDasharray: "4 4",
                      strokeWidth: "2",
                      d: "M107.36 2.48l105.7 185.47H2.66L108.35 2.48z",
                      clipRule: "evenodd"
                    }), /* @__PURE__ */jsx("ellipse", {
                      cx: "182.68",
                      cy: "156.48",
                      fill: "transparent",
                      rx: "74.32",
                      ry: "74.52"
                    }), /* @__PURE__ */jsx("path", {
                      stroke: scheme === "dark" ? "white" : "black",
                      strokeWidth: "2",
                      d: "M256.5 156.48c0 40.88-33.05 74.02-73.82 74.02-40.77 0-73.83-33.14-73.83-74.02 0-40.87 33.06-74.01 73.83-74.01 40.77 0 73.82 33.14 73.82 74.01z"
                    }), /* @__PURE__ */jsx("mask", {
                      id: "a",
                      width: "149",
                      height: "150",
                      x: "108",
                      y: "81",
                      maskUnits: "userSpaceOnUse",
                      children: /* @__PURE__ */jsx("ellipse", {
                        cx: "182.68",
                        cy: "156.48",
                        fill: "white",
                        rx: "74.32",
                        ry: "74.52"
                      })
                    }), /* @__PURE__ */jsx("g", {
                      mask: "url(#a)",
                      children: /* @__PURE__ */jsx("path", {
                        fill: scheme === "dark" ? "white" : "black",
                        fillRule: "evenodd",
                        d: "M108.36 2.48l105.7 185.47H2.66L108.35 2.48z",
                        clipRule: "evenodd"
                      })
                    })]
                  }), /* @__PURE__ */jsxs(Flex, {
                    direction: "column",
                    align: "center",
                    padding: 4,
                    children: [/* @__PURE__ */jsx(Text, {
                      size: 3,
                      children: "No deployments created yet."
                    }), /* @__PURE__ */jsx(Box, {
                      padding: 4,
                      children: /* @__PURE__ */jsx(Button, {
                        fontSize: 3,
                        paddingX: 5,
                        paddingY: 4,
                        tone: "primary",
                        radius: 4,
                        text: "Add Project",
                        onClick: () => setIsFormOpen(true)
                      })
                    }), /* @__PURE__ */jsx(Text, {
                      size: 1,
                      weight: "semibold",
                      muted: true,
                      children: /* @__PURE__ */jsx("a", {
                        href: "https://github.com/ndimatteo/sanity-plugin-vercel-deploy#-your-first-vercel-deployment",
                        target: "_blank",
                        rel: "noopener noreferrer",
                        style: {
                          color: "inherit"
                        },
                        children: "Need help?"
                      })
                    })]
                  })]
                })
              })
            })
          })]
        })
      }), isFormOpen && /* @__PURE__ */jsx(Dialog, {
        header: "New Project Deployment",
        id: "create-webhook",
        width: 1,
        onClickOutside: () => setIsFormOpen(false),
        onClose: () => setIsFormOpen(false),
        footer: /* @__PURE__ */jsx(Box, {
          padding: 3,
          children: /* @__PURE__ */jsxs(Grid, {
            columns: 2,
            gap: 3,
            children: [/* @__PURE__ */jsx(Button, {
              padding: 4,
              mode: "ghost",
              text: "Cancel",
              onClick: () => setIsFormOpen(false)
            }), /* @__PURE__ */jsx(Button, {
              padding: 4,
              text: "Create",
              tone: "primary",
              loading: isSubmitting,
              onClick: () => onSubmit(),
              disabled: isSubmitting || !pendingDeploy.project || !pendingDeploy.url || !pendingDeploy.token
            })]
          })
        }),
        children: /* @__PURE__ */jsx(Box, {
          padding: 4,
          children: /* @__PURE__ */jsxs(Stack, {
            space: 4,
            children: [/* @__PURE__ */jsx(FormField, {
              title: "Display Title (internal use only)",
              description: /* @__PURE__ */jsxs(Fragment, {
                children: ["This should be the environment you are deploying to, like", " ", /* @__PURE__ */jsx("em", {
                  children: "Production"
                }), " or ", /* @__PURE__ */jsx("em", {
                  children: "Staging"
                })]
              }),
              children: /* @__PURE__ */jsx(TextInput, {
                type: "text",
                value: pendingDeploy.title,
                onChange: e => {
                  e.persist();
                  const title = e.target.value;
                  setpendingDeploy(prevState => ({
                    ...prevState,
                    ...{
                      title
                    }
                  }));
                }
              })
            }), /* @__PURE__ */jsx(FormField, {
              title: "Vercel Project Name",
              description: "Vercel Project: Settings \u2192 General \u2192 \"Project Name\"",
              children: /* @__PURE__ */jsx(TextInput, {
                type: "text",
                value: pendingDeploy.project,
                onChange: e => {
                  e.persist();
                  const project = e.target.value;
                  setpendingDeploy(prevState => ({
                    ...prevState,
                    ...{
                      project
                    }
                  }));
                }
              })
            }), /* @__PURE__ */jsx(FormField, {
              title: "Vercel Team Name",
              description: "Required for projects under a Vercel Team: Settings \u2192 General \u2192 \"Team Name\"",
              children: /* @__PURE__ */jsx(TextInput, {
                type: "text",
                value: pendingDeploy.team,
                onChange: e => {
                  e.persist();
                  const team = e.target.value;
                  setpendingDeploy(prevState => ({
                    ...prevState,
                    ...{
                      team
                    }
                  }));
                }
              })
            }), /* @__PURE__ */jsx(FormField, {
              title: "Deploy Hook URL",
              description: "Vercel Project: Settings \u2192 Git \u2192 \"Deploy Hooks\"",
              children: /* @__PURE__ */jsx(TextInput, {
                type: "text",
                inputMode: "url",
                value: pendingDeploy.url,
                onChange: e => {
                  e.persist();
                  const url = e.target.value;
                  setpendingDeploy(prevState => ({
                    ...prevState,
                    ...{
                      url
                    }
                  }));
                }
              })
            }), /* @__PURE__ */jsx(FormField, {
              title: "Vercel Token",
              description: "Vercel Account dropdown: Settings \u2192 \"Tokens\"",
              children: /* @__PURE__ */jsx(TextInput, {
                type: "text",
                value: pendingDeploy.token,
                onChange: e => {
                  e.persist();
                  const token = e.target.value;
                  setpendingDeploy(prevState => ({
                    ...prevState,
                    ...{
                      token
                    }
                  }));
                }
              })
            }), /* @__PURE__ */jsx(FormField, {
              children: /* @__PURE__ */jsx(Card, {
                paddingY: 3,
                children: /* @__PURE__ */jsxs(Flex, {
                  align: "center",
                  children: [/* @__PURE__ */jsx(Switch, {
                    id: "disableDeleteAction",
                    style: {
                      display: "block"
                    },
                    onChange: e => {
                      e.persist();
                      const isChecked = e.target.checked;
                      setpendingDeploy(prevState => ({
                        ...prevState,
                        ...{
                          disableDeleteAction: isChecked
                        }
                      }));
                    },
                    checked: pendingDeploy.disableDeleteAction
                  }), /* @__PURE__ */jsx(Box, {
                    flex: 1,
                    paddingLeft: 3,
                    children: /* @__PURE__ */jsx(Text, {
                      children: /* @__PURE__ */jsx("label", {
                        htmlFor: "disableDeleteAction",
                        children: 'Disable the "Delete" action for this item in production?'
                      })
                    })
                  })]
                })
              })
            })]
          })
        })
      })]
    })
  });
};
const vercelDeployTool = definePlugin(options => {
  const {
    name,
    title,
    icon,
    ...config
  } = options || {};
  return {
    name: "sanity-plugin-vercel-deploy",
    tools: [{
      name: name || "vercel-deploy",
      title: title || "Deploy",
      icon: icon || deployIcon,
      component: VercelDeploy,
      options: config,
      router: route.create("/*")
    }]
  };
});
export { vercelDeployTool };
