// next-sanity/dist/studio/NextStudioNoScript.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var jsxRuntime = require('react/jsx-runtime');
const style = {
  __html: "\n.sanity-app-no-js__root {\n  position: absolute;\n  top: 0;\n  right: 0;\n  left: 0;\n  bottom: 0;\n  background: #fff;\n  z-index: 1;\n}\n\n.sanity-app-no-js__content {\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  transform: translate(-50%, -50%);\n  text-align: center;\n  font-family: helvetica, arial, sans-serif;\n}\n"
};
const NextStudioNoScript = () => /* @__PURE__ */jsxRuntime.jsx("noscript", {
  children: /* @__PURE__ */jsxRuntime.jsx("div", {
    className: "sanity-app-no-js__root",
    children: /* @__PURE__ */jsxRuntime.jsxs("div", {
      className: "sanity-app-no-js__content",
      children: [/* @__PURE__ */jsxRuntime.jsx("style", {
        type: "text/css",
        dangerouslySetInnerHTML: style
      }), /* @__PURE__ */jsxRuntime.jsx("h1", {
        children: "JavaScript disabled"
      }), /* @__PURE__ */jsxRuntime.jsxs("p", {
        children: ["Please ", /* @__PURE__ */jsxRuntime.jsx("a", {
          href: "https://www.enable-javascript.com/",
          children: "enable JavaScript"
        }), " in your browser and reload the page to proceed."]
      })]
    })
  })
});
exports.NextStudioNoScript = NextStudioNoScript;
