// @sanity/demo/dist/tailwind.js
import { black, white, hues } from '@sanity/color';
import { space, radius, fontFamily, screen } from '@sanity/demo';
function _buildTailwindColorTints(tints) {
  return Object.fromEntries(Object.entries(tints).map(_ref => {
    let [tintKey, tint] = _ref;
    return [tintKey, tint.hex];
  }));
}
const _round = num => num.toFixed(7).replace(/(\.[0-9]+?)0+$/, "$1").replace(/\.0$/, "");
const _rem = px => "".concat(_round(px / 16), "rem");
const theme = {
  extend: {
    spacing: {
      1: "".concat(space[1], "px"),
      2: "".concat(space[2], "px"),
      3: "".concat(space[3], "px"),
      4: "".concat(space[4], "px"),
      5: "".concat(space[5], "px"),
      6: "".concat(space[6], "px"),
      7: "".concat(space[7], "px"),
      8: "".concat(space[8], "px"),
      9: "".concat(space[9], "px")
    },
    typography: {
      DEFAULT: {
        css: {
          "--tw-prose-body": black.hex,
          "--tw-prose-headings": black.hex,
          "--tw-prose-lead": black.hex,
          "--tw-prose-links": black.hex,
          "--tw-prose-bold": black.hex,
          "--tw-prose-counters": black.hex,
          "--tw-prose-bullets": black.hex,
          "--tw-prose-hr": black.hex,
          "--tw-prose-quotes": black.hex,
          "--tw-prose-quote-borders": black.hex,
          "--tw-prose-captions": black.hex,
          "--tw-prose-code": black.hex,
          "--tw-prose-pre-code": white.hex,
          "--tw-prose-pre-bg": black.hex,
          "--tw-prose-th-borders": black.hex,
          "--tw-prose-td-borders": black.hex,
          "--tw-prose-invert-body": white.hex,
          "--tw-prose-invert-headings": white.hex,
          "--tw-prose-invert-lead": white.hex,
          "--tw-prose-invert-links": white.hex,
          "--tw-prose-invert-bold": white.hex,
          "--tw-prose-invert-counters": white.hex,
          "--tw-prose-invert-bullets": white.hex,
          "--tw-prose-invert-hr": white.hex,
          "--tw-prose-invert-quotes": white.hex,
          "--tw-prose-invert-quote-borders": white.hex,
          "--tw-prose-invert-captions": white.hex,
          "--tw-prose-invert-code": white.hex,
          "--tw-prose-invert-pre-code": black.hex,
          "--tw-prose-invert-pre-bg": white.hex,
          "--tw-prose-invert-th-borders": white.hex,
          "--tw-prose-invert-td-borders": white.hex,
          blockquote: {
            fontWeight: "inherit",
            fontStyle: "normal"
          }
        }
      },
      sm: {
        css: {
          pre: {
            borderRadius: _rem(radius[2])
          }
        }
      },
      base: {
        css: {
          pre: {
            borderRadius: _rem(radius[2])
          }
        }
      },
      lg: {
        css: {
          pre: {
            borderRadius: _rem(radius[2])
          }
        }
      },
      xl: {
        css: {
          pre: {
            borderRadius: _rem(radius[2])
          }
        }
      },
      "2xl": {
        css: {
          pre: {
            borderRadius: _rem(radius[2])
          }
        }
      }
    }
  },
  borderRadius: {
    none: "".concat(radius[0]),
    sm: "".concat(radius[1], "px"),
    DEFAULT: "".concat(radius[2], "px"),
    md: "".concat(radius[3], "px"),
    lg: "".concat(radius[4], "px"),
    full: "9999px"
  },
  colors: {
    transparent: "transparent",
    current: "currentColor",
    white: white.hex,
    black: black.hex,
    gray: _buildTailwindColorTints(hues.gray),
    blue: _buildTailwindColorTints(hues.blue),
    purple: _buildTailwindColorTints(hues.purple),
    magenta: _buildTailwindColorTints(hues.magenta),
    red: _buildTailwindColorTints(hues.red),
    orange: _buildTailwindColorTints(hues.orange),
    yellow: _buildTailwindColorTints(hues.yellow),
    green: _buildTailwindColorTints(hues.green),
    cyan: _buildTailwindColorTints(hues.cyan)
  },
  fontFamily,
  screens: {
    sm: "".concat(screen[0], "px"),
    md: "".concat(screen[1], "px"),
    lg: "".concat(screen[2], "px"),
    xl: "".concat(screen[3], "px"),
    "2xl": "".concat(screen[4], "px")
  }
};
export { theme };
